serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractDeserializer.java

```diff
@@ -0,0 +1,122 @@
+import io.pravega.schemaregistry.contract.data.EncodingId;
+import io.pravega.schemaregistry.contract.data.EncodingInfo;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.schemas.Schema;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+@Slf4j
+abstract class AbstractDeserializer<T> extends BaseDeserializer<T> {
+    private static final int HEADER_SIZE = 1 + Integer.BYTES;
+
+    private final String groupId;
+    private final SchemaRegistryClient client;
+    // This can be null. If no schema is supplied, it means the intent is to deserialize into writer schema. 
+    // If headers are not encoded, then this will be the latest schema from the registry
+    private final SchemaInfo schemaInfo;
+    private final boolean encodeHeader;
+    private final SerializerConfig.Decoders decoders;
+    private final boolean skipHeaders;
+    private final EncodingCache encodingCache;
+
+    protected AbstractDeserializer(String groupId,
+                                   SchemaRegistryClient client,
+                                   @Nullable Schema<T> schema,
+                                   boolean skipHeaders,
+                                   SerializerConfig.Decoders decoders,
+                                   EncodingCache encodingCache,
+                                   boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(encodingCache);
+        this.groupId = groupId;
+        this.client = client;
+        this.encodingCache = encodingCache;
+        this.schemaInfo = schema == null ? null : schema.getSchemaInfo();
+        this.encodeHeader = encodeHeader;
+        this.skipHeaders = skipHeaders;
+        this.decoders = decoders;
+            
+        initialize();
+    }
+
+    private void initialize() {
+        if (schemaInfo != null) {
+            log.info("Validate caller supplied schema.");
+            if (!client.canReadUsing(groupId, schemaInfo)) {
+                throw new IllegalArgumentException("Cannot read using schema" + schemaInfo.getType() + " as it is considered incompatible with current policy.");
+            }
+        } else {
+            if (!this.encodeHeader) {
+                log.warn("No reader schema is supplied and stream does not have encoding headers.");
+            }
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public T deserialize(ByteBuffer data) {
+        int start = data.hasArray() ? data.arrayOffset() + data.position() : data.position();
+        if (this.encodeHeader) {
+            SchemaInfo writerSchema = null;
+            ByteBuffer decoded;
+            if (skipHeaders) {
```

Comment:
I am not clear on why one would set encodeHeader to true and skipHeader to true as opposed to setting encodeHeader to false. It looks like in both cases it means that `writerSchema` is null.

Reply:
Skip header is only an optimization flag used during deserialization. 
It is used to tell the deserializer to skip the encoded header and not to talk to the service to resolve it to the write time schema.. instead use the supplied read time schema to read into. 

This is relevant for protobuf and json typed deserializers where user supplies a read time schema to read into.